### DISCUSSION TITLE
Ubuntu based systems hanging when running sunbird services locally

### DISCUSSION THREAD

**Discussion by Debayanpaul1593** [Sunbird-Ed/Community#601] [Sunbird ED configuration discussion] [answered]
Hi Sunbird Team,

We are facing a few challenges when running sunbird building blocks locally in our Ubuntu based systems. At any point of time we are running cassandra, neo4j, ES, redis, kafka/zookeeper, keycloak on docker(all/some of these services) and along with these sunbird lern and sunbird knowlg java services. We are running them on local machines for development purposes. However ubuntu systems are behaving irratically which is causing problems for the development team. Some of these problems are:
1. Erratic System crashes
2. System unresponsive
3. System lagging
4. System Overheating

PFA configuration of the system:
![Screenshot from 2023-08-01 14-09-57](https://github.com/Sunbird-Ed/Community/assets/37288058/d7d5d5e1-6bae-46f6-b365-fa2e62a384b7)

We are working on Mac systems as well but they are much more resilient to these issues. In calls we had earlier with the sunbird team we noticed many developers working with Ubuntu systems. If there is any particular specification which can help us reduce/resolve these issues - be it Ubuntu version/processor specification, we will try to get it fixed. Please inform us of the same. It will help us immensely.

Thank you

**Comment by gandham-santhosh** [comment]
Hi @Debayanpaul1593 , if I am understanding right, you are running all the services listed above on a single server in a standalone mode. This will definitely chock as these services may be consuming CPU/Memory more than available in the server. Considering the number of services you are running, I suggest you to distribute these services across multiple servers, java applications on one server, databases on other servers. 
BTW, are you following any instructions/documentation to deploy these services/databases to the server?
Also, I am just trying to understand your objective, are you trying to setup only Sunbird-Lern BB services and develop/customise on top of it?

**Reply by Debayanpaul1593** [reply]
We have deployed our databases/services on server on a much smaller scale for testing/development purposes only. This is because we are customising on top of certain microservices such as sunbird-lern and sunbird-knowlg and as such have not deployed using deployment scripts but by following the instructions given in repository README files for local deployment. 

When running certain Sunbird microservices such as Sunbird Lern User Org Service on local machines(for debugging purposes) we are connecting to database services deployed in our dev/test server. But there is still a lot of issues seen in ubuntu systems(but not seen in Macs). 

We were wondering if there are any issues with Ubuntu 22.04.2 LTS which are causing such erratic behaviour. If sunbird team is using Ubuntu for development then what is the OS version? Also what is the configuration of development machines that run Ubuntu?

**Reply by gandham-santhosh** [reply, marked as answer]
Most of our developers use MAC and some of them use Ubuntu 20.04 LTS as well with 8 Core 16 GB RAM preferably SSD disk. Multi-threading is handled well in MAC, that may be the reason you are not seeing the performance issues with MAC. 
Did you check the utilisation of the process that is running on Ubuntu using _ps_ or _htop_ commands? Can you also check if it is due to high consumption of CPU or Memory(RAM) ? If it is due to memory, you can reduce your maxHeap setting and run the process
Also, I suggest checking if there are any other processes running in the system apart from Sunbird microservices which might be consuming a lot of CPU and Memory
